tensorflow/contrib/rnn/python/ops/rnn_cell.py

```diff
@@ -3392,3 +3393,305 @@ def call(self, inputs, state):
+
+    self._init_read_vectors = [self.add_variable(
+        'initial_read_vector_%d' % i,
+        shape=[1, self.memory_vector_dim],
+        initializer=initializers.xavier_initializer())
+        for i in range(self.read_head_num)]
+
+    self._init_address_weights = [self.add_variable(
+        'initial_address_weights_%d' % i,
+        shape=[1, self.memory_size],
+        initializer=initializers.xavier_initializer())
+        for i in range(self.read_head_num + self.write_head_num)]
+
+    self.built = True
+
+  def call(self, x, prev_state):
+    # Addressing Mechanisms (Sec 3.3)
+
+    prev_read_vector_list = control_flow_ops.cond(
+      math_ops.equal(prev_state.time, 0),
+      lambda: [self._expand(
+        math_ops.tanh(
+          array_ops.squeeze(
+            math_ops.matmul(
+              array_ops.ones([1, 1]),
+              self._init_read_vectors[i]))),
+        dim=0, N=x.get_shape()[0])
+        for i in range(self.read_head_num)],
+      lambda: prev_state.read_vector_list)
+    if self.read_head_num == 1:
+      prev_read_vector_list = [prev_read_vector_list]
+
+    controller_input = array_ops.concat([x] + prev_read_vector_list, axis=1)
+    with vs.variable_scope('controller', reuse=self.reuse):
+      controller_output, controller_state = self.controller(controller_input,
+        prev_state.controller_state)
+
+    parameters = math_ops.matmul(controller_output, self._params_kernel)
+    parameters = nn_ops.bias_add(parameters, self._params_bias)
+    parameters = clip_ops.clip_by_value(parameters,
+      -self.clip_value, self.clip_value)
+    head_parameter_list = array_ops.split(
+      parameters[:, :self.num_parameters_per_head * self.num_heads],
+      self.num_heads,
+      axis=1)
+    erase_add_list = array_ops.split(
+      parameters[:, self.num_parameters_per_head * self.num_heads:],
+      2 * self.write_head_num,
+      axis=1)
+
+    prev_w_list = control_flow_ops.cond(
+      math_ops.equal(prev_state.time, 0),
+      lambda: [self._expand(
+        nn_ops.softmax(array_ops.squeeze(
+            math_ops.matmul(
+              array_ops.ones([1, 1]),
+              self._init_address_weights[i]))),
+        dim=0, N=x.get_shape()[0])
+        for i in range(self.read_head_num + self.write_head_num)],
+      lambda: prev_state.w_list)
+    if (self.read_head_num + self.write_head_num) == 1:
+      prev_w_list = [prev_w_list]
+
+    prev_M = prev_state.M
+    w_list = []
+    for i, head_parameter in enumerate(head_parameter_list):
+      k = math_ops.tanh(head_parameter[:, 0:self.memory_vector_dim])
+      beta = nn_ops.softplus(head_parameter[:, self.memory_vector_dim])
+      g = math_ops.sigmoid(head_parameter[:, self.memory_vector_dim + 1])
+      s = nn_ops.softmax(head_parameter[:,self.memory_vector_dim + 2:
+        self.memory_vector_dim + 2 + (self.shift_range * 2 + 1)])
+      gamma = nn_ops.softplus(head_parameter[:, -1]) + 1
+      with vs.variable_scope('addressing_head_%d' % i):
```

Comment:
can addressing access required variables via, e.g., self._addressing_head_variables[i]["kernel"] etc?  and these get built in `build`?